power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
+    std::array<std::tuple<Branch3Side, Branch3Side>, 3> const branch3_combinations{
+        {{side_1, side_2}, {side_2, side_3}, {side_1, side_3}}};
+    for (auto const& transformer3w : state.components.template citer<ThreeWindingTransformer>()) {
+        for (auto const& [first_side, second_side] : branch3_combinations) {
+            if (!transformer3w.status(first_side) || !transformer3w.status(second_side)) {
+                continue;
+            }
+            auto const& from_node = transformer3w.node(first_side);
+            auto const& to_node = transformer3w.node(second_side);
+
+            auto const tap_at_first_side = transformer3w.tap_side() == first_side;
+            auto const single_direction_condition = regulated_objects.transformers3w.contains(transformer3w.id()) &&
+                                                    (tap_at_first_side || transformer3w.tap_side() == second_side);
+            if (single_direction_condition) {
+                auto const& tap_side_node = tap_at_first_side ? from_node : to_node;
+                auto const& non_tap_side_node = tap_at_first_side ? to_node : from_node;
+                Idx2D const regulated_idx =
+                    assigned_idx ? unregulated_idx : main_core::get_component_idx_by_id(state, transformer3w.id());
+                assigned_idx = true;
+                add_to_edge(edges, edge_props, tap_side_node, non_tap_side_node, {regulated_idx, 1});
+            } else {
+                add_to_edge(edges, edge_props, from_node, to_node, {unregulated_idx, 1});
+                add_to_edge(edges, edge_props, to_node, from_node, {unregulated_idx, 1});
+            }
+        }
+    }
+}
+
+template <std::derived_from<Transformer> Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
+constexpr void add_edge(main_core::MainModelState<ComponentContainer> const& state,
+                        RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
+                        TrafoGraphEdgeProperties& edge_props) {
+    for (auto const& transformer : state.components.template citer<Transformer>()) {
+        if (!transformer.from_status() || !transformer.to_status()) {
+            continue;
+        }
+        auto const& from_node = transformer.from_node();
+        auto const& to_node = transformer.to_node();
+
+        if (regulated_objects.transformers.contains(transformer.id())) {
+            auto const tap_at_from_side = transformer.tap_side() == BranchSide::from;
+            auto const& tap_side_node = tap_at_from_side ? from_node : to_node;
+            auto const& non_tap_side_node = tap_at_from_side ? to_node : from_node;
+            if (get_component<Node>(state, tap_side_node).u_rated() <
+                get_component<Node>(state, non_tap_side_node).u_rated()) {
+                throw AutomaticTapCalculationError(transformer.id());
+            }
```

Comment:
shouldn't a similar check happen for 3w transformer? if so, is this really the right location? can't we do that at construction time of the 3w transformer? feels like a typical use case of `assert_valid_input_data`

Reply:
Its not in algorithm for 3 winding transformer. Like i proposed at parallel 0 weighted edge case, adding this condition can also address that practically unlikely edge case.

Nice idea though, if implemented at `assert_valid_input_data`, it should not matter if automatic tap changer gives divergence or incorrect result by not satisfying this condition.

Reply:
why doesn't it apply for 3w transformer? you still have the requirement that the u_rated at the tap side should be geq the u_rated at the regulated side, right? especially since it is modeled as pi-configuration and therefore effectively makes the following configuration
```
        /-oo- B
A -oo- D
        \-oo- C
```

if A is now regulated, then `u_rated(A) >= u_rated(D) >= u_rated(B), u_rated(C)`

otherwise, if the single-regulated-transformer configuration `A -oo- D` can give rise to divergences and errors, then so can the regulated 3w threewinding transformer